docs/adr/010_device_request_queue.md

```diff
@@ -0,0 +1,154 @@
+be summed up as follows:
+
+1. Add a queue to `LoRaDevice` for all device-related operations. There is one
+   that exists today but it is used to service uplink data requests only. The
+   idea would be to extend it to encompass all other requests and operations.
+
+1. Turn the LNS implementation to be (logically) single-threaded so that not
+   only does it make it easier to reason about the code (except perhaps with
+   regards to re-entrancy), but it also enables use of very simple data
+   structures without the need for locking. This is possible because the bulk
+   of LNS is I/O bound. There is practically no CPU-intensive code that
+   executes between each await.
+
+1. Simplify the entire flow and processing of a message into a simple
+   request-response model that can be manipulated and reasoned about much more
+   easily. That is, all methods receiving and processing a `LoRaRequest`
+   return `Task<DownlinkMessage>` instead of `void` or `Task`. This allows all
+   `Task.Run` uses to be moved to a single and central point when a message is
+   received instead of being littered throughout the code base. The main
+   message loop could also be made responsible for central
+   error-handling/logging and sending of downlink messages when the processing
+   of a request has completed. This would just require regular use of tasks.
+   Next, all `Task`-based operations on `LoRaDevice` could be naturally and
+   _implicitly_ queued and then executed in a mutually exclusive manner.
+
+## Decision
+
+All operations for a particular device requiring the IoT Hub connection are to
+be executed in a serial/exclusive fashion on the `LoRaDevice` itself. This
+will ensure that there is only one operation acting on the connection at any
+given time. Making each execution exclusive (think queue) gives a
+deterministic way of adding a close operation without affecting any other
+operations that may be in flight.
+
+Unfortunately, the initialization of a `LoRaDevice` cannot be addressed with
+the same approach since it requires a two-step initialization per its current
+design. That is, a `LoRaDevice` is _minimally initialized_ when created so its
+`InitializeAsync` method can be called to fetch the twin information. Until
+that completes, a `LoRaDevice` is not considered technically ready for being
+put in the cache. Since a `LoRaDevice` cannot be discovered during its
+initialization, it cannot be synchronized with the other operations. To work
+around this,m load operations will be synchronized in the `LoRaDeviceRegistry`
+to ensure there is only ever one operation loading the twins for a particular
+DevEUI. It was also decided that closing the connection after initialization
+is the easiest approach to delay the connection ownership decision to after
+the arrival of the first message.
+
+The approach to simplify the entire flow and processing of a message into a
+simple request-response model will be taken a stretch goal so that the
```

Comment:
```suggestion
simple request-response model will be taken as stretch goal so that the
```

Reply:
Thanks. Actually the fix here was that "as" was just missing so I added it with 80be0392ea9700fe185dd718f228b4ff1775fb32.